satellite/repair/checker/observer.go

```diff
@@ -0,0 +1,406 @@
+func (observer *RangedLoopObserver) getNodesEstimate(ctx context.Context) (int, error) {
+	// this should be safe to call frequently; it is an efficient caching lookup.
+	totalNumNodes, err := observer.nodestate.NumNodes(ctx)
+	if err != nil {
+		// We could proceed here by returning the last good value, or by returning a fallback
+		// constant estimate, like "20000", and we'd probably be fine, but it would be better
+		// not to have that happen silently for too long. Also, if we can't get this from the
+		// database, we probably can't modify the injured segments queue, so it won't help to
+		// proceed with this repair operation.
+		return 0, err
+	}
+
+	if totalNumNodes == 0 {
+		return 0, Error.New("segment health is meaningless: there are no nodes")
+	}
+
+	return totalNumNodes, nil
+}
+
+func (observer *RangedLoopObserver) createInsertBuffer() *queue.InsertBuffer {
+	return queue.NewInsertBuffer(observer.repairQueue, observer.repairQueueBatchSize)
+}
+
+// TestingCompareInjuredSegmentIDs compares stream id of injured segment.
+func (observer *RangedLoopObserver) TestingCompareInjuredSegmentIDs(ctx context.Context, streamIDs []uuid.UUID) error {
+	injuredSegments, err := observer.repairQueue.SelectN(ctx, 100)
+	if err != nil {
+		return err
+	}
+
+	var injuredSegmentsIds []uuid.UUID
+	for _, segment := range injuredSegments {
+		injuredSegmentsIds = append(injuredSegmentsIds, segment.StreamID)
+	}
+
+	sort.Slice(injuredSegmentsIds, func(i, j int) bool {
+		return injuredSegmentsIds[i].Less(injuredSegmentsIds[j])
+	})
+
+	sort.Slice(streamIDs, func(i, j int) bool {
+		return streamIDs[i].Less(streamIDs[j])
+	})
+
+	if !reflect.DeepEqual(streamIDs, injuredSegmentsIds) {
+		return errs.New("injured objects ids are different")
+	}
+
+	return nil
+}
+
+// Start starts parallel segments loop.
+func (observer *RangedLoopObserver) Start(ctx context.Context, startTime time.Time) error {
+	observer.counter1.Reset() // reset counter values before new ranged loop run
+	observer.counter2.Reset()
+	observer.counter3.Reset()
+	observer.counter4.Reset()
+	observer.counter5.Reset()
+
+	observer.TotalStats = aggregateStats{}
+	observer.startTime = startTime
+	return nil
+}
+
+// Fork creates a Partial to process a chunk of all the segments.
+func (observer *RangedLoopObserver) Fork(ctx context.Context) (rangedloop.Partial, error) {
+	return newRangedLoopCheckerPartial(observer), nil
+}
+
+// Join is called after the chunk for Partial is done.
+// This gives the opportunity to merge the output like in a reduce step.
+func (observer *RangedLoopObserver) Join(ctx context.Context, partial rangedloop.Partial) error {
+	repPartial, ok := partial.(*repairPartial)
+	if !ok {
+		return Error.New("expected partial type %T but got %T", repPartial, partial)
+	}
+
+	if err := repPartial.repairQueue.Flush(ctx); err != nil {
+		return Error.Wrap(err)
+	}
+
+	observer.combineStats(repPartial.monStats)
+	observer.statsCollector.aggregateStats(repPartial.statsCollector)
```

Comment:
I think we're getting close. This function still emits some observations that are for the partial only that I _think_ are intended to cover the whole run. Instead of aggregating into statsCollector here, what if the observer also maintained a map[string]*batchStats that was reset on each run that we aggregated the stats gathered by the partial into.

Then, once, in Finish, we can push that final, aggregated map[string]*batchStats into the stats collector....